File: android/app/src/main/java/com/ayuvo/health/insights/HealthAgeEngine.kt
package com.ayuvo.health.insights

import com.ayuvo.health.insights.InsightsMath.clamp
import com.ayuvo.health.insights.InsightsMath.interp
import com.ayuvo.health.insights.InsightsMath.mean
import com.ayuvo.health.insights.InsightsMath.roundTo
import com.ayuvo.health.insights.InsightsMath.sampleSd
import com.ayuvo.health.insights.InsightsMath.sleepMidpointMin
import java.time.LocalDate

/**
 * Ayuvo Health Age (docs/insights.md, "How Ayuvo Health Age is estimated"): Ayuvo's own estimate,
 * not a clinical or biological age. Each marker's 90-day average becomes a capped years offset;
 * the weighted mean of the available offsets (capped at ±10) is added to the actual age. Ported
 * from `_marker`, `_health_age_raw`, `health_age` and `health_age_pace`.
 */
object HealthAgeEngine {

    private class MarkerValue(val value: Double?, val secondary: Double?, val basis: String?, val days: Int, val equivalentAge: Double?, val offset: Double?)

    private fun sexTable(tables: Map<String, Points>, sex: String?): Points {
        if (sex == "male" || sex == "female") return tables.getValue(sex)
        val m = tables.getValue("male")
        val f = tables.getValue("female")
        return m.indices.map { i -> m[i].first to (m[i].second + f[i].second) / 2.0 }
    }

    private fun window(series: Map<LocalDate, Double>, asOf: LocalDate, days: Int): List<Double> =
        InsightsMath.values(series, asOf, -(days - 1), 0).map { it.second }

    private fun marker(mk: InsightsConfig.Marker, inputs: InsightsInputs, asOf: LocalDate, profile: InsightsProfile, actual: Double, cfg: InsightsConfig): MarkerValue {
        val ha = cfg.healthAge
        val win = ha.windowDays
        val sex = profile.sex
        return when (mk.method) {
            "age_norm", "dose_response" -> {
                val vals = window(inputs.series(mk.series!!), asOf, win)
                if (vals.size < mk.minDays) return MarkerValue(null, null, null, vals.size, null, null)
                val v = mean(vals)
                if (mk.method == "dose_response") return MarkerValue(v, null, null, vals.size, null, interp(mk.points, v))
                val tables = if (mk.tablesByKind.isNotEmpty()) mk.tablesByKind.getValue(inputs.hrvKind) else mk.tables
                val eq = InsightsMath.inverseAge(sexTable(tables, sex), v, ha.ageMin, ha.ageMax)
                MarkerValue(v, null, null, vals.size, eq, eq - actual)
            }
            "sleep" -> {
                val nights = BaselineEngine.validNights(inputs, cfg)
                val use = (win - 1 downTo 0).map { asOf.minusDays(it.toLong()) }.filter { it in nights }
                if (use.size < mk.minDays) return MarkerValue(null, null, null, use.size, null, null)
                val hours = mean(use.map { nights.getValue(it).asleepMin!! / 60.0 })
                val mids = use.map { sleepMidpointMin(nights.getValue(it), it, inputs.timeZone).toDouble() }
                val sd = sampleSd(mids, mean(mids))
                MarkerValue(hours, sd, null, use.size, null, interp(mk.durationPoints, hours) + interp(mk.regularityPoints, sd))
            }
            "workouts" -> {
                val wear = window(inputs.series("steps"), asOf, win).size
                if (!inputs.tracking.workouts || wear < mk.minDays) return MarkerValue(null, null, null, wear, null, null)
                val loads = TrainingLoad.dailyLoads(inputs.workouts, inputs.timeZone, cfg)
                var total = 0.0
                var active = 0
                for (w in mk.weeks - 1 downTo 0) {
                    var block = 0.0
                    for (k in 6 downTo 0) block += loads[asOf.minusDays((7 * w + k).toLong())]?.minutes ?: 0.0
                    total += block
                    if (block > 0) active++
                }
                val perWeek = total / mk.weeks
                val share = active / mk.weeks.toDouble()
                MarkerValue(perWeek, share, null, wear, null, interp(mk.minutesPoints, perWeek) + interp(mk.activeSharePoints, share))
            }
            "body_composition" -> {
                val fat = window(inputs.series("body_fat"), asOf, win)
                if ((sex == "male" || sex == "female") && fat.size >= mk.minDays) {
                    val v = mean(fat)
                    return MarkerValue(v, null, "body_fat", fat.size, null, interp(mk.bodyFatPoints.getValue(sex), v))
                }
                var bmis = window(inputs.series("bmi"), asOf, win)
                val height = profile.heightCm
                if (bmis.isEmpty() && height != null && height != 0.0) {
                    val h = height / 100.0
                    bmis = window(inputs.series("weight"), asOf, win).map { it / (h * h) }
                }
                if (bmis.size < mk.minDays) return MarkerValue(null, null, null, bmis.size, null, null)
                val v = mean(bmis)
                MarkerValue(v, null, "bmi", bmis.size, null, interp(mk.bmiPoints, v))
            }
            else -> error("unknown marker method ${mk.method}")
        }
    }

    /** Unrounded result; [HealthAgeResult] fields hold full precision here. */
    private fun raw(inputs: InsightsInputs, asOf: LocalDate, profile: InsightsProfile, cfg: InsightsConfig): HealthAgeResult {
        val ha = cfg.healthAge
        fun out(status: String, actual: Double? = null, markers: List<HealthAgeMarker> = emptyList(), collecting: Collecting? = null, available: Int = 0) =
            HealthAgeResult(asOf, status, actual, null, null, null, markers, collecting, available, ha.minMarkers)
        val birthday = profile.birthday ?: return out("no_birthday")
        val actual = InsightsMath.daysBetween(birthday, asOf) / ha.daysPerYear
        if (actual < ha.minActualAge) return out("unsupported_age", actual)
        var wsum = 0.0
        val markers = ArrayList<HealthAgeMarker>()
        for (mk in ha.markers) {
            val v = marker(mk, inputs, asOf, profile, actual, cfg)
            val offset = v.offset?.let { clamp(it, -mk.capYears, mk.capYears) }
            if (offset != null) wsum += mk.weight
            markers += HealthAgeMarker(
                id = mk.id, method = mk.method, available = offset != null, value = v.value, secondaryValue = v.secondary,
                basis = v.basis, days = v.days, neededDays = mk.minDays, equivalentAge = v.equivalentAge, offsetYears = offset,
                weight = mk.weight, contributionYears = null
            )
        }
        val avail = markers.filter { it.available }
        val core = ha.coreMarkers.toSet()
        if (avail.size < ha.minMarkers || avail.none { it.id in core }) {
            val need = ha.collectingDays
            val best = (markers.filter { it.id in core }.map { it.days } + 0).max()
            return out("collecting", actual, markers, Collecting(minOf(best, need), need), avail.size)
        }
        var acc = 0.0
        val withContribution = markers.map { m ->
            if (!m.available) m else {
                val c = m.weight * m.offsetYears!! / wsum
                acc += c
                m.copy(contributionYears = c)
            }
        }
        val diff = clamp(acc, -ha.totalCapYears, ha.totalCapYears)
        val ids = avail.map { it.id }.toSet()
        val confidence = when {
            avail.size >= ha.highMinMarkers && ha.highRequires in ids -> "high"
            avail.size >= ha.mediumMinMarkers -> "medium"
            else -> "low"
        }
        return HealthAgeResult(asOf, "ok", actual, actual + diff, diff, confidence, withContribution, null, avail.size, ha.minMarkers)
    }

    /** Ayuvo Health Age as of [asOf] (inclusive 90-day window). */
    fun healthAge(inputs: InsightsInputs, asOf: LocalDate, profile: InsightsProfile, cfg: InsightsConfig): HealthAgeResult {
        val r = raw(inputs, asOf, profile, cfg)
        return r.copy(
            actualAge = roundTo(r.actualAge, 1),
            healthAge = roundTo(r.healthAge, 1),
            difference = roundTo(r.difference, 1),
            markers = r.markers.map {
                it.copy(
                    value = roundTo(it.value, 2), secondaryValue = roundTo(it.secondaryValue, 2), equivalentAge = roundTo(it.equivalentAge, 1),
                    offsetYears = roundTo(it.offsetYears, 2), contributionYears = roundTo(it.contributionYears, 2)
                )
            }
        )
    }

    /**
     * Health Age recomputed as of each of the last `weeks` week-ends (as of, as of − 7 … as of −
     * 7·weeks); pace is the least-squares slope in years of Health Age per calendar year.
     */
    fun pace(inputs: InsightsInputs, asOf: LocalDate, profile: InsightsProfile, cfg: InsightsConfig): HealthAgePace {
        val p = cfg.healthAge
        val pts = ArrayList<Triple<Int, LocalDate, HealthAgeResult>>()
        for (k in p.paceWeeks downTo 0) {
            val d = asOf.minusDays(7L * k)
            val r = raw(inputs, d, profile, cfg)
            if (r.ok) pts += Triple(-7 * k, d, r)
        }
        val points = pts.map { (_, d, r) -> HealthAgePoint(d, roundTo(r.healthAge!!, 1), roundTo(r.difference!!, 1)) }
        if (pts.size < p.paceMinPoints) return HealthAgePace("insufficient", pts.size, p.paceMinPoints, null, null, points)
        val slope = InsightsMath.olsSlope(pts.map { (x, _, r) -> x.toDouble() to r.healthAge!! })!!
        val pace = slope * p.daysPerYear
        val direction = if (pace < p.paceImprovingBelow) "improving" else if (pace > p.paceDecliningAbove) "declining" else "stable"
        return HealthAgePace("ok", pts.size, p.paceMinPoints, roundTo(pace, 2), direction, points)
    }
}

File: ios/calorietracker/Insights/HealthAgeEngine.swift
import Foundation

// Ayuvo Health Age: each marker's 90-day average turned into a capped years offset, weighted over the available
// markers, plus a 12-week pace. Port of the "Health Age" section of `scripts/insights_reference.py`.

nonisolated enum HealthAgeEngine {
    private struct MarkerValue {
        var value: Double?
        var secondary: Double?
        var basis: String?
        var days: Int
        var equivalentAge: Double?
        var offset: Double?
    }

    /// Sex-specific table; any other or missing sex → the element-wise mean of the male and female tables.
    private static func sexTable(_ tables: [String: [[Double]]], _ sex: String?) -> [[Double]] {
        if let sex, sex == "male" || sex == "female", let t = tables[sex] { return t }
        let m = tables["male"] ?? [], f = tables["female"] ?? []
        return zip(m, f).map { [$0[0], ($0[1] + $1[1]) / 2.0] }
    }

    private static func window(_ series: [String: Double], _ asOf: String, _ days: Int) -> [Double] {
        BaselineEngine.values(series, asOf, -(days - 1), 0).map(\.1)
    }

    private static func marker(_ mk: InsightsConfig.HealthAge.Marker, _ inputs: InsightsInputs, _ asOf: String,
                               _ profile: InsightsProfile, _ actual: Double, _ config: InsightsConfig) -> MarkerValue {
        let ha = config.healthAge
        let win = ha.windowDays
        let tz = InsightsDay.timeZone(inputs.timeZone)
        let sex = profile.sex
        switch mk.method {
        case "age_norm", "dose_response":
            let vals = window(inputs.series[mk.series ?? mk.id] ?? [:], asOf, win)
            if vals.count < mk.minDays { return MarkerValue(days: vals.count) }
            let v = InsightsMath.mean(vals)
            if mk.method == "dose_response" {
                return MarkerValue(value: v, days: vals.count, offset: InsightsMath.interp(mk.points ?? [[0, 0]], v))
            }
            let tables = mk.tablesByKind.map { $0[inputs.hrvKind] ?? [:] } ?? mk.tables ?? [:]
            let eq = InsightsMath.inverseAge(sexTable(tables, sex), v, ha.ageMin, ha.ageMax)
            return MarkerValue(value: v, days: vals.count, equivalentAge: eq, offset: eq - actual)
        case "sleep":
            let nights = BaselineEngine.validNights(inputs, config: config)
            var use: [(String, InsightsNight)] = []
            for k in stride(from: win - 1, through: 0, by: -1) {
                let d = InsightsDay.add(asOf, -k)
                if let n = nights[d] { use.append((d, n)) }
            }
            if use.count < mk.minDays { return MarkerValue(days: use.count) }
            let hours = InsightsMath.mean(use.map { $0.1.asleepMin / 60.0 })
            let mids = use.map { Double(InsightsDay.sleepMidpointMinutes(startMs: $0.1.startMs, endMs: $0.1.endMs, wakeDay: $0.0, tz)) }
            let sd = InsightsMath.sampleSD(mids, InsightsMath.mean(mids))
            return MarkerValue(value: hours, secondary: sd, days: use.count,
                               offset: InsightsMath.interp(mk.durationPoints ?? [[0, 0]], hours)
                                   + InsightsMath.interp(mk.regularityPoints ?? [[0, 0]], sd))
        case "workouts":
            let wear = window(inputs.series["steps"] ?? [:], asOf, win).count
            if !inputs.tracking.workouts || wear < mk.minDays { return MarkerValue(days: wear) }
            let loads = TrainingLoadEngine.dailyLoads(inputs.workouts, timeZone: tz, config: config)
            let weeks = mk.weeks ?? 12
            var total = 0.0
            var active = 0
            for w in stride(from: weeks - 1, through: 0, by: -1) {
                var block = 0.0
                for k in stride(from: 6, through: 0, by: -1) {
                    block += loads[InsightsDay.add(asOf, -(7 * w + k))]?.minutes ?? 0.0
                }
                total += block
                if block > 0 { active += 1 }
            }
            let perWeek = total / Double(weeks)
            let share = Double(active) / Double(weeks)
            return MarkerValue(value: perWeek, secondary: share, days: wear,
                               offset: InsightsMath.interp(mk.minutesPoints ?? [[0, 0]], perWeek)
                                   + InsightsMath.interp(mk.activeSharePoints ?? [[0, 0]], share))
        case "body_composition":
            let fat = window(inputs.series["body_fat"] ?? [:], asOf, win)
            if let sex, sex == "male" || sex == "female", fat.count >= mk.minDays, let points = mk.bodyFatPoints?[sex] {
                let v = InsightsMath.mean(fat)
                return MarkerValue(value: v, days: fat.count, offset: InsightsMath.interp(points, v))
            }
            var bmis = window(inputs.series["bmi"] ?? [:], asOf, win)
            if bmis.isEmpty, let height = profile.heightCm, height != 0 {
                let h = height / 100.0
                bmis = window(inputs.series["weight"] ?? [:], asOf, win).map { $0 / (h * h) }
            }
            if bmis.count < mk.minDays { return MarkerValue(days: bmis.count) }
            let v = InsightsMath.mean(bmis)
            return MarkerValue(value: v, basis: "bmi", days: bmis.count, offset: InsightsMath.interp(mk.bmiPoints ?? [[0, 0]], v))
        default:
            return MarkerValue(days: 0)
        }
    }

    /// Unrounded Health Age as of `asOf`.
    static func healthAgeRaw(_ inputs: InsightsInputs, asOf: String, profile: InsightsProfile, config: InsightsConfig) -> HealthAgeResult {
        let ha = config.healthAge
        var out = HealthAgeResult(asOf: asOf, status: "ok", markers: [], markersAvailable: 0, markersNeeded: ha.minMarkers)
        guard let bday = profile.birthday, !bday.isEmpty else {
            out.status = "no_birthday"
            return out
        }
        let actual = Double(InsightsDay.between(bday, asOf)) / ha.daysPerYear
        out.actualAge = actual
        if actual < ha.minActualAge {
            out.status = "unsupported_age"
            return out
        }
        var wsum = 0.0
        for mk in ha.markers {
            let r = marker(mk, inputs, asOf, profile, actual, config)
            var item = HealthAgeMarker(
                id: mk.id, method: mk.method, available: r.offset != nil, value: r.value, secondaryValue: r.secondary,
                basis: r.basis, days: r.days, neededDays: mk.minDays, equivalentAge: r.equivalentAge, offsetYears: nil,
                weight: mk.weight, contributionYears: nil)
            if mk.method == "body_composition", r.offset != nil, item.basis == nil { item.basis = "body_fat" }
            if let off = r.offset {
                item.offsetYears = InsightsMath.clamp(off, -mk.capYears, mk.capYears)
                wsum += mk.weight
            }
            out.markers.append(item)
        }
        let avail = out.markers.filter(\.available)
        out.markersAvailable = avail.count
        let core = Set(ha.coreMarkers)
        if avail.count < ha.minMarkers || !avail.contains(where: { core.contains($0.id) }) {
            let need = ha.collectingDays
            let best = (out.markers.filter { core.contains($0.id) }.map(\.days) + [0]).max() ?? 0
            out.status = "collecting"
            out.collecting = InsightsCollecting(have: min(best, need), need: need)
            return out
        }
        var acc = 0.0
        for idx in out.markers.indices where out.markers[idx].available {
            let c = out.markers[idx].weight * (out.markers[idx].offsetYears ?? 0) / wsum
            out.markers[idx].contributionYears = c
            acc += c
        }
        let diff = InsightsMath.clamp(acc, -ha.totalCapYears, ha.totalCapYears)
        let ids = Set(avail.map(\.id))
        let c = ha.confidence
        let conf: String
        if avail.count >= c.highMinMarkers, ids.contains(c.highRequires) {
            conf = "high"
        } else if avail.count >= c.mediumMinMarkers {
            conf = "medium"
        } else {
            conf = "low"
        }
        out.healthAge = actual + diff
        out.difference = diff
        out.confidence = conf
        return out
    }

    /// Ayuvo Health Age as of `asOf` (inclusive 90-day window).
    static func healthAge(_ inputs: InsightsInputs, asOf: String, profile: InsightsProfile, config: InsightsConfig) -> HealthAgeResult {
        var r = healthAgeRaw(inputs, asOf: asOf, profile: profile, config: config)
        r.actualAge = InsightsMath.roundTo(r.actualAge, 1)
        r.healthAge = InsightsMath.roundTo(r.healthAge, 1)
        r.difference = InsightsMath.roundTo(r.difference, 1)
        for idx in r.markers.indices {
            r.markers[idx].value = InsightsMath.roundTo(r.markers[idx].value, 2)
            r.markers[idx].secondaryValue = InsightsMath.roundTo(r.markers[idx].secondaryValue, 2)
            r.markers[idx].equivalentAge = InsightsMath.roundTo(r.markers[idx].equivalentAge, 1)
            r.markers[idx].offsetYears = InsightsMath.roundTo(r.markers[idx].offsetYears, 2)
            r.markers[idx].contributionYears = InsightsMath.roundTo(r.markers[idx].contributionYears, 2)
        }
        return r
    }

    /// Health Age as of each of the last `weeks` week-ends; pace = least-squares slope in years per calendar year.
    static func pace(_ inputs: InsightsInputs, asOf: String, profile: InsightsProfile, config: InsightsConfig) -> HealthAgePaceResult {
        let p = config.healthAge.pace
        var pts: [(Double, String, HealthAgeResult)] = []
        for k in stride(from: p.weeks, through: 0, by: -1) {
            let d = InsightsDay.add(asOf, -7 * k)
            let r = healthAgeRaw(inputs, asOf: d, profile: profile, config: config)
            if r.status == "ok" { pts.append((Double(-7 * k), d, r)) }
        }
        var out = HealthAgePaceResult(
            status: "insufficient", have: pts.count, needed: p.minPoints,
            points: pts.map { HealthAgePacePoint(day: $0.1, healthAge: InsightsMath.roundTo($0.2.healthAge, 1),
                                                 difference: InsightsMath.roundTo($0.2.difference, 1)) })
        guard pts.count >= p.minPoints,
              let slope = InsightsMath.olsSlope(pts.map { ($0.0, $0.2.healthAge ?? 0) }) else { return out }
        let pace = slope * config.healthAge.daysPerYear
        out.status = "ok"
        out.pace = InsightsMath.roundTo(pace, 2)
        out.direction = pace < p.improvingBelow ? "improving" : (pace > p.decliningAbove ? "declining" : "stable")
        return out
    }
}
